internal/http/services/thumbnails/thumbnails.go

```diff
@@ -0,0 +1,221 @@
+// Copyright 2018-2019 CERN
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// In applying this license, CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+package thumbnails
+
+import (
+	"bytes"
+	"crypto/md5"
+	"encoding/hex"
+	"errors"
+	"image"
+	"image/jpeg"
+	"image/png"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"os"
+	"path"
+	"strconv"
+
+	"github.com/cs3org/reva/pkg/appctx"
+	"github.com/cs3org/reva/pkg/rhttp"
+	"github.com/nfnt/resize"
+	"github.com/rs/zerolog"
+)
+
+type thumnailhandler struct {
+	Parent  *svc
+	Log     *zerolog.Logger
+	Writer  http.ResponseWriter
+	Request *http.Request
+	URL     string
+
+	FileType string
+	Size     uint
+
+	TempFileName string
+}
+
+func newThumbnailHandler(s *svc, w http.ResponseWriter, r *http.Request) *thumnailhandler {
+	t := thumnailhandler{Parent: s, Log: appctx.GetLogger(r.Context()), Writer: w, Request: r}
+
+	var u *url.URL
+	if s.conf.WebDavHost == "" {
+		u = &url.URL{}
+		u.Scheme = "http" // from where should that come
+		u.Host = r.Host
+	} else {
+		var err error
+		u, err = url.Parse(s.conf.WebDavHost)
+		if err != nil {
+			t.Log.Err(err).Msg("Failed to partse host: " + s.conf.WebDavHost)
+			t.Writer.WriteHeader(http.StatusInternalServerError)
+			return nil
```

Comment:
For the time being it does the job, but I wonder if it'd be possible not having this side effects, or decoupling the side effects from `t` and instead of mutating `t` along the body of the function, instead return a [composite struct literal](https://golang.org/ref/spec#Composite_literals) with the fields filled. It makes the logic easier to follow.

Reply:
One way of doing so is doing logging at the callers and return the error, this way `t` shouldn't need a logger.